Framework/API/test/FunctionAttributeValidatorTest.h

```diff
@@ -0,0 +1,216 @@
+#pragma once
+
+#include "MantidAPI/FunctionFactory.h"
+#include "MantidAPI/IFunction1D.h"
+#include "MantidKernel/ArrayBoundedValidator.h"
+#include "MantidKernel/BoundedValidator.h"
+#include "MantidKernel/ListValidator.h"
+#include "MantidKernel/StringContainsValidator.h"
+
+#include <cxxtest/TestSuite.h>
+
+using namespace Mantid;
+using namespace Mantid::API;
+
+namespace detail {
+class FAVT_Funct : public ParamFunction, public IFunction1D {
+public:
+  FAVT_Funct() {
+    declareAttribute("DAttr", Attribute(0.0), Mantid::Kernel::BoundedValidator<double>(0.0, 100.0));
+
+    declareAttribute("IAttr", Attribute(5), Mantid::Kernel::BoundedValidator<int>(0, 10));
+
+    declareAttribute("SAttr", Attribute("K"),
+                     Mantid::Kernel::StringListValidator(std::vector<std::string>{"K", "meV"}));
+
+    std::vector<std::string> sV(3);
+    sV[0] = "a";
+    sV[1] = "b";
+    sV[2] = "c";
+    declareAttribute("SQAttr", Attribute("abc", true), Mantid::Kernel::StringContainsValidator(sV));
+
+    std::vector<double> v(3);
+    v[0] = 1;
+    v[1] = 2;
+    v[2] = 3;
+    declareAttribute("VAttr", Attribute(v), Mantid::Kernel::ArrayBoundedValidator<double>(1, 5));
+
+    testInvalidDeclaration();
+  }
+
+  std::string name() const override { return "IFT_Funct"; }
+  void function1D(double *, const double *, const size_t) const override {}
+  void functionDeriv1D(Jacobian *, const double *, const size_t) override {}
+
+private:
+  void testInvalidDeclaration() {
+    TS_ASSERT_THROWS(
+        declareAttribute("DAttr_invalid", Attribute(-1), Mantid::Kernel::BoundedValidator<double>(0.0, 100.0)),
+        const IFunction::validationException &);
+
+    TS_ASSERT_THROWS(declareAttribute("IAttr_invalid", Attribute(11), Mantid::Kernel::BoundedValidator<int>(0, 10)),
+                     const IFunction::validationException &);
+
+    TS_ASSERT_THROWS(declareAttribute("SAttr_invalid", Attribute("Invalid"),
+                                      Mantid::Kernel::StringListValidator(std::vector<std::string>{"K", "meV"})),
+                     const IFunction::validationException &);
+
+    std::vector<std::string> sV(3);
+    sV[0] = "a";
+    sV[1] = "b";
+    sV[2] = "c";
+    TS_ASSERT_THROWS(
+        declareAttribute("SQAttr_invalid", Attribute("ab", true), Mantid::Kernel::StringContainsValidator(sV)),
+        const IFunction::validationException &);
+
+    std::vector<double> v(3);
+    v[0] = 1.0;
+    v[1] = 2.0;
+    v[2] = 50.0;
+    TS_ASSERT_THROWS(
+        declareAttribute("VAttr_invalid", Attribute(v), Mantid::Kernel::ArrayBoundedValidator<double>(1, 5)),
+        const IFunction::validationException &);
+  }
+};
+
+// Simple Attribute Visitor for Test
+class SetAttribute : public Mantid::API::IFunction::AttributeVisitor<> {
+public:
+  SetAttribute(Mantid::Kernel::IValidator_sptr validator) : m_dbl(0) { m_validator = validator; }
+  double m_dbl;
+  std::string m_str;
+  int m_i;
+  bool m_b;
```

Comment:
Ditto on uninitialized variables (ignore std::string as that default inits to "")